include/zephyr/sensing/sensing_sensor.h

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright (c) 2022-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SENSING_SENSOR_H_
+#define ZEPHYR_INCLUDE_SENSING_SENSOR_H_
+
+#include <zephyr/sensing/sensing.h>
+#include <zephyr/device.h>
+#include <stdbool.h>
+#include <zephyr/drivers/sensor.h>
+
+/**
+ * @defgroup sensing_sensor Sensor API
+ * @ingroup sensing
+ * @defgroup sensing_sensor_callbacks Sensor Callbacks
+ * @ingroup sensing_sensor
+ */
+
+/**
+ * @brief Sensor API
+ * @addtogroup sensing_sensor
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Sensor registration information
+ *
+ */
+struct sensing_sensor_register_info {
+
+	/**
+	 * Sensor flags
+	 */
+	uint16_t flags;
+
+	/**
+	 * Sample size in bytes for a single sample of the registered sensor.
+	 * sensing runtime need this information for internal buffer allocation.
+	 */
+	uint16_t sample_size;
+
+	/**
+	 * The number of sensor sensitivities
+	 */
+	uint8_t sensitivity_count;
+
+	/**
+	 * Sensor version.
+	 * Version can be used to identify different versions of sensor implementation.
+	 */
+	struct sensing_sensor_version version;
+};
+
+/**
+ * @brief Sensor context data structure
+ *
+ */
+struct sensing_sensor_ctx {
+
+	/**
+	 * For sensing runtime internal private data, sensor should not see and touch
+	 */
+	void *priv_ptr;
+
+	/**
+	 * Pointer to the sensor register information.
+	 */
+	const struct sensing_sensor_register_info *register_info;
+
+	/**
+	 * For sensor private context data, registered by sensor with \ref SENSING_SENSOR_DT_DEFINE.
+	 * Sensor could use \ref sensing_sensor_get_ctx_data to fetch out this filed with
+	 * struct device.
+	 */
+	void *const sensor_ctx_ptr;
+};
+
+static inline int sensing_sensor_dev_init(
+		const struct device *dev)
+{
+	/**
+	 * Nothing need to do in system auto initialization.
+	 * Sensor subsystem runtime will call each sensor instance's initialization
+	 * function via API callback according sensor reporting dependency sequences.
+	 * Sensor subsystem can make sure the depends sensor instances always initialized before
+	 * client sensors.
+	 */
+	return 0;
+}
+
+/**
+ * @brief Macro for define a sensor instance from devicetree node id
+ *
+ * This macro also defined a struct device for this sensor instance, and registered sensors'
+ * private context data, configuration data structure and API.
+ *
+ * sensing_init will enumerate all sensor instances from devicetree, and initialize each sensor
+ * instance defined by this macro.
+ *
+ */
+
+#define SENSING_SENSOR_DT_DEFINE(node_id, reg_ptr, ctx_ptr, api_ptr)			\
+	static struct sensing_sensor_ctx							\
+		_CONCAT(__sensing_sensor_ctx_, Z_DEVICE_DT_DEV_ID(node_id)) = {		\
+			.register_info = reg_ptr,					\
+			.sensor_ctx_ptr = ctx_ptr,					\
+		};									\
+	SENSOR_DEVICE_DT_DEFINE(node_id, sensing_sensor_dev_init, NULL,			\
```

Comment:
What's the reasoning behind this? If `CONFIG_SENSOR_INFO` is enabled, this will generate extra nodes in the sensor info array which will mess up the sensor shell.

Reply:
both SENSOR_DEVICE_DT_DEFINE and CONFIG_SENSOR_INFO  will pointer to the same device-tree variable, so it will not mess up with sensor shell.